replay events to client upon connection

ffakenz opened the pull request:
Contributes to #580 

🧃 Added new PersistenceClient handle to `loadAll` and `append` server output events.

🧃 Move Persistence handles to their own module.

To check before merging:
* [x] CHANGELOG is up to date
* [x] ~Documentation is up to date~

pgrange committed:
Prepare a test

v0d1ch committed:
Introduce Persistence module

ffakenz committed:
Make it compile

pass the persistence handler to the server

ch1bo committed:
Load and save server outputs to replay history

ch1bo committed:
Integrate the replay events assumption to the other E2E tests

v0d1ch committed:
CheckPoint - introduce append function in persistence

v0d1ch committed:
Implement append function in persistence

ffakenz committed:
Add new persistence client handle

pgrange committed:
Warning about a lot of file copying

v0d1ch committed:
Append newline on append and break the file contents on loadAll

ch1bo commented on hydra-node/src/Hydra/Persistence.hs:
```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE DuplicateRecordFields #-}
```
This should not be needed here?

ch1bo commented on hydra-node/src/Hydra/Persistence.hs:
```diff
@@ -0,0 +1,81 @@
+  }
+
+-- | Handle to append and load files to/from disk using JSON encoding.
+data PersistenceClient a m = PersistenceClient
```
I see we have a separate handle now for append semantics. Can we name this something like `AppendPersistence` or `PersistenceList` (not sure what this has to do with clients).

ch1bo commented on hydra-node/src/Hydra/Persistence.hs:
```diff
@@ -0,0 +1,81 @@
+                else do
+                  -- If 'a' already contains newlines then our persistence breaks
+                  let eresults = forM (C8.lines bs) $ \t -> Aeson.eitherDecodeStrict' t
+                  case eresults of
```
Could: write shorter: `forM (BS8.lines bs) Aeson.eitherDecodeStrict' >>= \case`

ch1bo commented on hydra-node/src/Hydra/Persistence.hs:
```diff
@@ -0,0 +1,81 @@
+      , append = \a -> do
+          -- REVIEW: what happens if the `a` here already contains the newline character? It would break all persistence!
+          -- Use DB?
+          let bytes = toStrict $ Aeson.encode a <> "\n"
```
Seems like aeson is using https://hackage.haskell.org/package/text-2.0.1/docs/Data-Text-Encoding.html#v:encodeUtf8BuilderEscaped to do the escaping. So this _should_ be fine.

Should: write a test though

pgrange replied:
Property test added.

ch1bo commented on hydra-node/src/Hydra/Persistence.hs:
```diff
@@ -0,0 +1,81 @@
+          -- Use DB?
+          let bytes = toStrict $ Aeson.encode a <> "\n"
+          -- atomicity in file writing implies a file copy everytime we append something to it
+          liftIO $ withBinaryFileDurableAtomic fp AppendMode (`BS.hPut` bytes)
```
Not doing the file moving ("atomic" party of this) is a valid concern here!

Should: remove `Atomic`

ch1bo replied:
We could have a look at the benchmarks to see how much impact this is having!

pgrange replied:
Not sure it is very convincing but here is what I see on my machine

With _Atomic_:
```
#> cabal bench hydra-cluster --benchmark-options "--scaling-factor 2"
...
Average confirmation time: 0.13254963166666667
```

Without:
```
Average confirmation time: 0.10909455041111112
```

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -30,61 +31,69 @@ spec :: Spec
-              Left{} -> failure $ "Failed to decode greeting " <> show received
+      withFreePort $ \port -> do
+        withTempDir "server-spec-tmp-dir" $ \tmp -> do
+          apiPersistence <- createPersistenceClient Proxy $ tmp <> "/server-output"
```
None of these tests is requiring persistence so we could just implement a `mockPersistence` which does not save/load antyhing.

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-unused-do-bind #-}
```
Should: remove this.

ch1bo reviewed (changes requested):
Some minor things still left. Most importantly test the Persistence handles as they become more involved.

ch1bo committed:
Rename PersistenceClient -> PersistenceIncremental

ch1bo committed:
Remove unused pragmas and imports

ch1bo committed:
Start a persistence test

pgrange committed:
Progressing on the test

v0d1ch committed:
Introduce simple generators

ffakenz committed:
Start using monadicIO again

pgrange committed:
Integrate withTempDir with property test

pgrange committed:
Starting to label properties

ch1bo committed:
Check coverage in persistence property test

pgrange committed:
Expect at least 10% of newLines in tests

ch1bo committed:
Not copy whole file content on append in PersistenceIncremental

By not requiring the write to be atomic we can avoid copying.

ch1bo committed:
Add some more Persistence tests

ch1bo committed:
Change createPersistence to not take a Proxy

We actually don't need it on the construction of the Persistence handle,
but call sites of load/save will be sufficient for inferring the type.

ch1bo committed:
Do not use real persistence in ServerSpec tests

ch1bo committed:
Update changelog

ch1bo reviewed (approved):
This is in good shape now, but we still need to do a bit more work to get #580 done.

ch1bo committed:
Send Greetings on server start

See note and api.yaml why we need to do this and it's consequence of
treating all peers as disconnected.

ch1bo committed:
Add a note in Changelog about the Greetings message

ch1bo committed:
Also mention replay of api outputs in documentation
